# ML lab 01 - TensorFlow의 설치및 기본적인 operations (new)

## Installing Tensorflow

```sh
pip install --upgrade tensorflow
```

## Check installation and version

In [1]:
import tensorflow as tf
print(tf.__version__)

1.1.0


## TensorFlow Hello World!

In [2]:
# Create a constant op
# This op is added as a node ato the default graph
hello = tf.constant("Hello, TensorFlow!")

# seart a TF session
sess = tf.Session()

# run the op and get result
print(sess.run(hello))

b'Hello, TensorFlow!'


## Computational Graph

In [3]:
# Building graph
node1 = tf.constant(3.0, tf.float32)
node2 = tf.constant(4.0)  # also tf.float32 implictily
node3 = tf.add(node1, node2)

In [4]:
print("node1:", node1, "node2:", node2)
print("node3:", node3)

node1: Tensor("Const_1:0", shape=(), dtype=float32) node2: Tensor("Const_2:0", shape=(), dtype=float32)
node3: Tensor("Add:0", shape=(), dtype=float32)


In [5]:
# feed data and run graph
sess = tf.Session()
print("sess.run(node1, node2): ", sess.run([node1, node2]))
print("sess.run(node3): ", sess.run(node3))

sess.run(node1, node2):  [3.0, 4.0]
sess.run(node3):  7.0


## Placeholder

In [6]:
a = tf.placeholder(tf.float32)
b = tf.placeholder(tf.float32)
adder_node = a + b # + provides a shorcut for tf.add(a, b)

print(sess.run(adder_node, feed_dict={a: 3, b: 4.5}))
print(sess.run(adder_node, feed_dict={a: [1, 3], b: [2, 4]}))

7.5
[ 3.  7.]


## Tensor Ranks, Shapes, and Types

### Rank

| Rank | Math entity | Python example |
| :--: | :---------: | :------------: |
| 0 | Scalar(magnitude only | s = 483 |
| 1 | Vector(magnitude and dirction) | v = [1.1, 2.2, 3.3] |
| 2 | Matrix(table of numbers) | m = [[1, 2, 3], [4, 5, 6], [7, 8, 9]] |
| 3 | 3-Tensor(cube of number) | t = [[[2], [4], [6]], [[8], [10], [12]], [[14], [16], [18]]]
| n | n-Tensor(you get the idea) | .... |

### Shape

| Rank | Shape | Dimension number | Example |
| :--: | :---: | :--------------: | :-----: |
| 0 | [] | 0-D | A 0-D tensor. A scalar. |
| 1 | [D0] | 1-D | A 1-D tensor with shape [5]. |
| 2 | [D0, D1] | 2-D | A 2-D tensor with shape [3, 4]. |
| 3 | [D0, D1, D2] | 3-D | A 3-D tensor with shape [1, 4, 3]. |
| n | [D0, D1, ..., Dn-1] | n-D | A tensor with shape [D0, D1, ..., Dn-1]. |

### Types

| Data type | Python type | Description |
| :-------: | :---------: | :---------: |
| **DT_FLOAT** | **tf.float32** | **32 bits floating point.** |
| DT_DOUBLE | tf.float64 | 64 bits floating point. |
| DT_INT8 | tf.int8 | 8 bits signed integer. |
| DT_INT16 | tf.int16 | 16 bits signed integer. |
| DT_INT32 | tf.int32 | 32 bits signed integer. |
| DT_INT64 | tf.int64 | 64 bits signed integer. |
| ... | ... | ... |